src/code/providers/provider-interface.ts

```diff
@@ -0,0 +1,550 @@
+// Sanity-check the conversion and remove this comment.
+/*
+ * decaffeinate suggestions:
+ * DS201: Simplify complex destructure assignments
+ * DS203: Remove `|| {}` from converted for-own loops
+ * DS205: Consider reworking code to avoid use of IIFEs
+ * DS207: Consider shorter variations of null checks
+ * Full docs: https://github.com/decaffeinate/decaffeinate/blob/master/docs/suggestions.md
+ */
+
+import isString  from '../utils/is-string'
+import _ from 'lodash'
+
+const FILE_EXTENSION_DELIMETER = "."
+
+type providerType = string
+type parentType = string // Based on the code, seems like this might path?
+type providerDataType = any  // Not sure what this is
+type documentId = string
+type key = string
+type mimeType = string
+
+
+// TODO: What does a save callback signature really looklike?
+type ISaveCallback = (arg: any) => any
+
+// TODO: What does a load callback signature really looklike?
+type ILoadCallback = (arg: any) => any
+
+// TODO: What does a list callback signature really looklike?
+type IListCallback = (arg: any) => any
+
+// TODO: What does a rename callback signature really looklike?
+type IRenameCallback = (arg: any) => any
+
+// TODO: What does a list callback signature really looklike?
+type IRemoveCallback = (arg: any) => any
+
+// TODO: What does a close callback signature really looklike?
+type ICloseCallback = (arg: any) => any
+
+// TODO: What does a open-save callback signature really looklike?
+type IOpenSaveCallback = (arg: any) => any
+
+
+enum ICloudFileTypes {
+  File = "file",
+  Folder = "folder",
+  Label = "label",
+  Extension = "extension"
+}
+
+interface ICloudMetaDataOpts {
+  name: string
+  description: string
+  content: any
+  url: string
+  type: ICloudFileTypes
+  provider: providerType
+  parent: ICloudMetaDataOpts|null
+  providerData: providerDataType
+  overwritable: boolean
+  sharedContentId: documentId
+  sharedContentSecretKey: key
+  mimeType: mimeType
+  ReadableExtensions?: string[]
+}
+interface ISharedMetadata {
+  _permissions?: any,
+  shareEditKey?: string,
+  sharedDocumentId?: string,
+  accessKeys?: string | string[]
+}
+
+type CloudFileContentType = any
+interface ICloudFileOpts {
+  content: CloudFileContentType
+  metadata: ICloudMetaDataOpts
+}
+
+class CloudFile {
+  content: CloudFileContentType
+  metadata: ICloudMetaDataOpts
+  constructor(options: ICloudFileOpts) {
+    ({content: this.content, metadata: this.metadata} = options)
+  }
+}
+
+class CloudMetadata {
+  name: string
+  description: string
+  content: any
+  url: string
+  type: ICloudFileTypes
+  provider: providerType
+  parent: ICloudMetaDataOpts
+  providerData: providerDataType
+  overwritable: boolean
+  sharedContentId: documentId
+  sharedContentSecretKey: key
+  mimeType: mimeType
+  filename: string|null
+
+  static Folder = 'folder'
+  static File = 'file'
+  static Label = 'label'
+
+  // TODO IMPORTANT: These are defined as class variables,
```

Comment:
Yeah, I'm not sure.  I think it will become obvious once you start typing the providers that use this.